invoice/invoice.go

```diff
@@ -0,0 +1,966 @@
+	}
+
+	if invoice.FallbackAddr != nil {
+		var version byte
+		switch addr := invoice.FallbackAddr.(type) {
+		case *btcutil.AddressPubKeyHash:
+			version = 17
+		case *btcutil.AddressScriptHash:
+			version = 18
+		case *btcutil.AddressWitnessPubKeyHash:
+			version = addr.WitnessVersion()
+		case *btcutil.AddressWitnessScriptHash:
+			version = addr.WitnessVersion()
+		default:
+			return fmt.Errorf("unknown fallback address type")
+		}
+		base32Addr, err := bech32.ConvertBits(
+			invoice.FallbackAddr.ScriptAddress(), 8, 5, true)
+		if err != nil {
+			return err
+		}
+
+		writeTaggedField(bufferBase32, fieldTypeF,
+			append([]byte{version}, base32Addr...))
+	}
+
+	if len(invoice.RoutingInfo) > 0 {
+		// Each extra routing info is encoded using 51 bytes.
+		routingDataBase256 := make([]byte, 0, 51*len(invoice.RoutingInfo))
+		for _, r := range invoice.RoutingInfo {
+			base256 := make([]byte, 51)
+			copy(base256[:33], r.PubKey.SerializeCompressed())
+			binary.BigEndian.PutUint64(base256[33:41], r.ShortChanID)
+			binary.BigEndian.PutUint64(base256[41:49], r.Fee)
+			binary.BigEndian.PutUint16(base256[49:51], r.CltvExpDelta)
+			routingDataBase256 = append(routingDataBase256, base256...)
+		}
+		routingDataBase32, err := bech32.ConvertBits(routingDataBase256,
+			8, 5, true)
+		if err != nil {
+			return err
+		}
+
+		writeTaggedField(bufferBase32, fieldTypeR, routingDataBase32)
+	}
+
+	if invoice.PubKey != nil {
+		// Convert 33 byte pubkey to 53 5-bit groups.
+		pubKeyBase32, err := bech32.ConvertBits(
+			invoice.PubKey.SerializeCompressed(), 8, 5, true)
+		if err != nil {
+			return nil
+		}
+
+		if len(pubKeyBase32) != 53 {
+			return fmt.Errorf("invalid pubkey length: %d",
+				len(invoice.PubKey.SerializeCompressed()))
+		}
+
+		writeTaggedField(bufferBase32, fieldTypeN, pubKeyBase32)
+	}
+
+	return nil
+}
+
+// writeTaggedField takes the type of a tagged data field, and the data of
+// the tagged field (encoded in base32), and writes the type, length and data
+// to the buffer.
+func writeTaggedField(bufferBase32 *bytes.Buffer, dataType byte, data []byte) error {
+	// Length must be exactly 10 bits, so add leading zero groups if
+	// needed.
+	lenBase32 := uint64ToBase32(uint64(len(data)))
+	for len(lenBase32) < 2 {
+		lenBase32 = append([]byte{0}, lenBase32...)
+	}
+
+	if len(lenBase32) != 2 {
+		return fmt.Errorf("data length too big to fit within 10 bits: %d",
+			len(data))
+	}
+
+	bufferBase32.WriteByte(dataType)
+	bufferBase32.Write(lenBase32)
+	bufferBase32.Write(data)
+
+	return nil
+}
+
+// decodeAmount returns the amount encoded by the provided string in
+// milllisatoshi.
+func decodeAmount(amount string) (uint64, error) {
+	if len(amount) < 1 {
+		return 0, fmt.Errorf("amount must be non-empty")
+	}
+
+	// If last character is a digit, then the amount can just be
+	// interpreted as BTC.
+	char := amount[len(amount)-1]
+	digit := char - '0'
+	if digit >= 0 && digit <= 9 {
+		btc, err := strconv.ParseUint(amount, 10, 64)
```

Comment:
If BTC can be specified instead of satoshis, then we can't use a regular `ParseUint` here, right?

Reply:
Not sure why we cannot.. It will just be a number?